File: commands/stats.js
const Func        = require('../libs/functions.js');

let options = {
    weekday: "long", year: "numeric", month: "short",
    day: "numeric", hour: "2-digit", minute: "2-digit"
};

module.exports = function(Discord,request,Config,message) {
    if (Func.isCommand(message, '!stats') || Func.isCommand(message, '!s')) {
        if (!Func.getParams(message)) return message.channel.reply("**SYNTAX:** !stats <username>");

        request.get('http://typeracerdata.com/api?username=' + Func.getParams(message)[1], function(error, response, body) {    
            if (error || !body || response.statusCode != 200) return message.channel.reply("The username you have entered could not be found, please try again!");
            data = JSON.parse(body);

            /* Shouldn't have to add this, but we'll try */
            if (!data) return message.channel.reply("Uh oh, there is c")
            
            let last = new Date(data.account.last_import * 1000);
            let marathon = new Date(data.account.marathon_start * 1000);

            const startEmbed = new Discord.RichEmbed()
                .setColor('#00B2EE')
                .setURL('http://www.typeracerdata.com/profile?username=' + data.account.username)
                .setThumbnail(`http://data.typeracer.com/misc/pic?uid=tr:` + data.account.username + `&size=large&bpc=1`)
                .setAuthor(`Statistics for ` + data.account.username)
                .setDescription(`
                    The data supplied by this bot is from TypeRacerData's API which was created by Noah. Make sure to check out the TypeRacer Championship at http://typeracerdata.com/championship/
                `)
                .addField('\u200b', '__**General Statistics**__', true)
                .addField('Races:', data.account.races + ` (` + data.account.wins + ` won)`)
                .addField('Texts:', data.account.texts_raced)
                .addField('Marathon:', data.account.marathon_total + ` on ` + marathon.toLocaleTimeString("en-us", options))
                .addField('\u200b', '__**WPM Statistics**__', true)
                .addField('Career Avg:', parseFloat(data.account.wpm_life).toFixed(2) + ` WPM`)
                .addField('Highest:', parseFloat(data.account.wpm_highest).toFixed(2) + ` WPM`)
                .addField('Text Bests:', parseFloat(data.account.wpm_textbests).toFixed(2) + ` WPM`)
                .addField('Last 10:', parseFloat(data.account.wpm_last10).toFixed(2) + ` WPM ` + `(` + parseFloat(data.account.wpm_bestlast10).toFixed(2) + ` peak)`)
                .setFooter(`Last Imported: ` + last.toLocaleTimeString("en-us", options))
            ;

            try { 
                message.channel.sendEmbed(startEmbed);
            } catch (err) {
                message.channel.reply('There was a problem, please try again!');
            }
        });
    }
}

File: discord.py/cmds/stats.py
"""
    File: stats.py
    Description: Executes the command !s or !stats
    Last Modified: 6/4/2017
"""
# pylint: disable=C0301

import datetime
import json
import urllib.request
import discord
from methods import send_reply

async def execute(client, message, config):
    """ Executes the command !stats """

    if config['Commands']['stats'] is True or message.channel.permissions_for(message.author).kick_members or message.author.id == config['Owner']:
        args = message.content.split(' ')

        if len(args) is 1:
            await send_reply(client, message.channel, message.author.id, 'Please use the correct syntax. !stats <username>', True)
        else:
            try:
                response = urllib.request.urlopen('http://typeracerdata.com/api?username=' + args[1])
                data = json.load(response)
            except ValueError:
                await send_reply(client, message.channel, message.author.id, 'the username you have entered does not exist, please try again!')

            if data:
                timezone = 'UTC'
                marathon = datetime.datetime.fromtimestamp(float(data['account']['marathon_start'])).strftime('%Y-%m-%d %H:%M:%S') + ' ' + timezone
                lastimport = datetime.datetime.fromtimestamp(float(data['account']['last_import'])).strftime('%Y-%m-%d %H:%M:%S') + ' ' + timezone

                embed = discord.Embed(colour=0x00e5ee)
                embed.set_author(name='Statistics for ' + args[1], icon_url='http://data.typeracer.com/misc/pic?uid=tr:' + args[1] + '&size=large&bpc=1')
                embed.set_thumbnail(url='http://data.typeracer.com/misc/pic?uid=tr:' + args[1] + '&size=large&bpc=1')
                embed.add_field(name='\u200b', value='__**General Statistics**__', inline=False)
                embed.add_field(name='Races:', value=data['account']['races'] + ' (' + data['account']['wins'] + ' won)')
                embed.add_field(name='Texts:', value=data['account']['texts_raced'])
                embed.add_field(name='Marathon:', value=data['account']['marathon_total'] + ' races on ' + marathon)
                embed.add_field(name='\u200b', value='__**WPM Statistics**__', inline=False)
                embed.add_field(name='Career Avg:', value='%.2f' % float(data['account']['wpm_life']) + ' WPM')
                embed.add_field(name='Highest:', value='%.2f' % float(data['account']['wpm_highest']) + ' WPM')
                embed.add_field(name='Text Bests:', value='%.2f' % float(data['account']['wpm_textbests']) + ' WPM')
                embed.add_field(name='Last 10:', value='%.2f' % float(data['account']['wpm_last10']) + ' WPM ' + '(' + '%.2f' % float(data['account']['wpm_bestlast10']) + ' peak)')
                embed.set_footer(text='Updated on ' + lastimport)

                await client.send_typing(message.channel)
                await client.send_message(message.channel, embed=embed)
    else:
        await client.send_typing(message.channel)
        await send_reply(client, message.channel, message.author.id, 'this command has been temporarily disabled!')
